Feature/docker management

trashstackdev opened the pull request:
# Pull Request

## Trello Ticket

- [Trello Ticket Link](insert link here)
  <!-- Link to the corresponding Trello ticket for better context -->

## Description

<!-- Provide a brief description of the changes introduced by this pull request. -->

## Additional Remarks

<!-- Any additional comments or context that would be helpful for the reviewer. -->

krieger2501 committed:
implement create posts endpoint

trashstackdev committed:
Add server_alpha_db SQL deployment script

This commit introduces the SQL deployment script for the server_alpha_db. The script contains code for creating the database, schemas, tables, and their related constraints. It also includes user roles and access configurations.

trashstackdev committed:
Refactor scripts and update README

The commit refactors several scripts, introduces a new 'clean_db.sh' script, and deletes the 'clean.sh' script. Moreover, the README has been updated with a 'Getting Started' section. These changes improve script usability, clarity and make running, building, and cleaning tasks easier and more intuitive.

trashstackdev committed:
Update JWT key handling and add docker configurations

Refactored the JWT key handling in the jwt_manager.go to fetch keys from environment variables rather than a file on disk. This increases the security of the JWT keys. The initializer.go was also updated to handle missing database environment variables more robustly, preventing server crashes when these variables aren't set. Dockerfile and docker-compose.yml have been added and updated for better containerization of the application. This includes non-root user creation in Dockerfile for improved security.

trashstackdev committed:
Update JWT key handling and add docker configurations

Refactored the JWT key handling in the jwt_manager.go to fetch keys from environment variables rather than a file on disk. This increases the security of the JWT keys. The initializer.go was also updated to handle missing database environment variables more robustly, preventing server crashes when these variables aren't set. Dockerfile and docker-compose.yml have been added and updated for better containerization of the application. This includes non-root user creation in Dockerfile for improved security.

trashstackdev committed:
Merge remote-tracking branch 'origin/feature/docker-management' into feature/docker-management

# Conflicts:
#	internal/initializer.go

krieger2501 commented on build/package/Dockerfile:
```diff
@@ -14,8 +14,17 @@ RUN go build -o server-alpha ./cmd/server
 
 ## Serve only the compiled binary in the second stage
 FROM alpine:3.17.6 AS serve
+WORKDIR /go/app
```
Wofür das hier? Würde das jetzt im Endeffekt nicht den späteren Copy-Befehl in /go/app/go/app setzen?

krieger2501 commented on deployments/docker-compose.yml:
```diff
@@ -1 +1,30 @@
+      - ./server_alpha_db.sql:/docker-entrypoint-initdb.d/initdb.sql
+
+  app:
+    image: "server-alpha:latest"
```
Könnten wir theoretisch auch vom GitHub-Container Registry ziehen, oder?

krieger2501 commented on deployments/server_alpha_db.sql:
```diff
@@ -0,0 +1,187 @@
+
+-- object: alpha_schema | type: SCHEMA --
+-- DROP SCHEMA IF EXISTS alpha_schema CASCADE;
+CREATE SCHEMA alpha_schema;
```
Schema muss v1 heißen, sonst funktioniert der gesamte Code nicht. Können es aber auch im Code renamen, wenn es dir so lieber ist

krieger2501 commented on deployments/server_alpha_db.sql:
```diff
@@ -0,0 +1,187 @@
+	password varchar(32) NOT NULL,
+	created_at date NOT NULL,
+	activated_at date,
+	expires_at date,
```
Wir benutzen kein Date, das ist zu ungenau, da es nur den Tag datiert. Um ganz genau zu sein wie abgesprochen in der Vorlesung, brauchen wir timestamptz, also nen Timestamp mit Timezone. Unten im Schema dann genauso

krieger2501 commented on internal/managers/jwt_manager.go:
```diff
@@ -1,61 +1,100 @@
 // NewJWTManager creates a new JWTManager with the initial key pair.
 func NewJWTManager() (JWTMgr, error) {
-	path := os.Getenv("KEY_PAIR_PATH")
+	privateKey, err := parsePrivateKey(os.Getenv("JWT_PRIVATE_KEY"))
+	if err != nil {
+		return nil, err
+	}
+	publicKey, err := parsePublicKey(os.Getenv("JWT_PUBLIC_KEY"))
+	if err != nil {
+		return nil, err
+	}
```
Ich versteh nicht ganz wieso du das alles jetzt geändert hast, gabs einen Grund dafür?

trashstackdev committed:
Update SQL queries to target 'alpha_schema'

The commit revises all SQL query strings in both 'post_handler.go' and 'user_handler.go'. It adjusts the targeted schema in these queries from 'v1' to 'alpha_schema'. This change aligns with modifications in our database structure where tables now reside in 'alpha_schema'.

trashstackdev committed:
Update date fields to timestamptz in SQL schemas

The date fields in the SQL schemas have been changed to timestamp with time zone (timestamptz). This update will allow more accurate tracking of timestamps across different timezones. It impacts the user, user_token, posts, subscriptions, and likes tables.

trashstackdev committed:
Add docker-compose.prod.yml and refine Docker configurations

Added a new docker-compose config file for prod environment. Additionally, refined configurations within the existing docker-compose and Dockerfile. Changes involved parameterization of environment variables to create a more flexible setup and adding a healthcheck for the database service. Removed unnecessary WORKDIR command from Dockerfile.

trashstackdev committed:
Update environment variable paths and names

Paths to environment (.env) files are updated to be more accurate across docker-compose files. Also, the naming of database related environment variables in docker-compose files as well as in initializer.go has been unified by removing the "LOCAL_" prefix to improve clarity.

trashstackdev committed:
Add script for environment setup and update .gitignore

Added script `setup_env.sh` to generate a `.env` file and add environment variables with default values. Based on this, the `.gitignore` file was also updated to include changes related to the `.env` file and key files. Additionally, `run.sh` was modified to check for the existence of a `.env` file and run `setup_env.sh` if not found.

trashstackdev committed:
Update `scripts/README.md` file

Updated and expanded the `scripts/README.md` file to reflect the purpose of the scripts, usage details and the environment they are intended for. Added guidance for using `run.sh` commands for building, starting, stopping services and cleaning the database.

trashstackdev committed:
Refine docker-compose commands and healthcheck script

The docker-compose commands in `run.sh` now use an environment file, and the healthcheck test in `docker-compose.yml` has been adjusted for specificity. This enhances script readability and Docker service control, while the updated healthcheck allows better PostgreSQL readiness validation.

trashstackdev committed:
Add APP_PORT variable and healthcheck endpoint, update docker-compose

Implemented an environment variable for APP_PORT, introduced a '/health' endpoint for the database healthcheck, and updated the docker-compose.yml configuration for the new port and healthcheck. This change aids in easier configuration and ensures more precise health assessments of the involved services.

krieger2501 commented on README.md:
```diff
@@ -1,2 +1,5 @@
 # server-alpha
 Server Repository for the Alpha Group (1)
+
+# Getting Started
```
Schreiben wir dann die Tage zusammen das HowTo?

trashstackdev replied:
Jo

krieger2501 commented on deployments/docker-compose.prod.yml:
```diff
@@ -0,0 +1,40 @@
+
+  app:
+    # It's possible to specify the image to use in the environment, default is the latest image from the github container registry
+    image: "${SERVER_IMAGE:-ghcr.io/wwi21seb-projekt/server-alpha:main}"
```
Funktioniert das oder brauchen wir nen Personal Access Token mit Rechten?

trashstackdev replied:
`docker pull ghcr.io/wwi21seb-projekt/server-alpha:main` klappt auf meinem Server ohne Access Tokens, denke also nicht

krieger2501 reviewed (approved):
`setup_env` ist nice, ty dafür :)

trashstackdev committed:
Update environment setup and deployment workflow

Introduced an environment variable `APP_HOST` and specified `ENVIRONMENT` for both development and production setups. Updated database health check and added one for the application in the production docker-compose.yml. Changed the password column in the server_alpha_db.sql to accommodate longer hashes. Integrated test execution in the CI workflow, build script and also added caching for Go modules. Modified CD pipeline to pull and deploy image on remote server.
